Generic type erased after type guard with strictNullChecks

**TypeScript Version:**  2.0.2

**Code**

```ts
interface Pet {
    name: string;
}

function isPet(pet: any): pet is Pet {
    return typeof pet.name === "string";
}

export function speak<TPet extends Pet>(pet: TPet, voice: (pet: TPet) => string): string {
    if (!isPet(pet)) {
        throw new Error("Expected \"pet\" to be a Pet");
    }

    return voice(pet); // line 14
}
```

**Expected behavior:**
Module compiles without error when `--strictNullChecks` is enabled OR disabled.

**Actual behavior:** 
- Module compiles without error when `--strictNullChecks` is _disabled_.
- The following error is produced when `--strictNullChecks` is _enabled_

    ```
    bug.ts(14,18): error TS2345: Argument of type 'Pet' is not assignable to parameter of type 'TPet'.
    ```

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -8682,7 +8682,7 @@ namespace ts {
                 // type. Otherwise, the types are completely unrelated, so narrow to an intersection of the
                 // two types.
                 const targetType = type.flags & TypeFlags.TypeParameter ? getApparentType(type) : type;
-                return isTypeSubtypeOf(candidate, targetType) ? candidate :
+                return isTypeSubtypeOf(candidate, type) ? candidate :
                     isTypeAssignableTo(type, candidate) ? type :
                     isTypeAssignableTo(candidate, targetType) ? candidate :
                     getIntersectionType([type, candidate]);

--- a/tests/baselines/reference/narrowingConstrainedTypeParameter.js
+++ b/tests/baselines/reference/narrowingConstrainedTypeParameter.js
@@ -1,0 +1,32 @@
+//// [narrowingConstrainedTypeParameter.ts]
+
+// Repro from #10811
+
+interface Pet {
+    name: string;
+}
+
+function isPet(pet: any): pet is Pet {
+    return typeof pet.name === "string";
+}
+
+export function speak<TPet extends Pet>(pet: TPet, voice: (pet: TPet) => string): string {
+    if (!isPet(pet)) {
+        throw new Error("Expected \"pet\" to be a Pet");
+    }
+    return voice(pet);
+}
+
+//// [narrowingConstrainedTypeParameter.js]
+// Repro from #10811
+"use strict";
+function isPet(pet) {
+    return typeof pet.name === "string";
+}
+function speak(pet, voice) {
+    if (!isPet(pet)) {
+        throw new Error("Expected \"pet\" to be a Pet");
+    }
+    return voice(pet);
+}
+exports.speak = speak;

--- a/tests/baselines/reference/narrowingConstrainedTypeParameter.symbols
+++ b/tests/baselines/reference/narrowingConstrainedTypeParameter.symbols
@@ -1,0 +1,42 @@
+=== tests/cases/compiler/narrowingConstrainedTypeParameter.ts ===
+
+// Repro from #10811
+
+interface Pet {
+>Pet : Symbol(Pet, Decl(narrowingConstrainedTypeParameter.ts, 0, 0))
+
+    name: string;
+>name : Symbol(Pet.name, Decl(narrowingConstrainedTypeParameter.ts, 3, 15))
+}
+
+function isPet(pet: any): pet is Pet {
+>isPet : Symbol(isPet, Decl(narrowingConstrainedTypeParameter.ts, 5, 1))
+>pet : Symbol(pet, Decl(narrowingConstrainedTypeParameter.ts, 7, 15))
+>pet : Symbol(pet, Decl(narrowingConstrainedTypeParameter.ts, 7, 15))
+>Pet : Symbol(Pet, Decl(narrowingConstrainedTypeParameter.ts, 0, 0))
+
+    return typeof pet.name === "string";
+>pet : Symbol(pet, Decl(narrowingConstrainedTypeParameter.ts, 7, 15))
+}
+
+export function speak<TPet extends Pet>(pet: TPet, voice: (pet: TPet) => string): string {
+>speak : Symbol(speak, Decl(narrowingConstrainedTypeParameter.ts, 9, 1))
+>TPet : Symbol(TPet, Decl(narrowingConstrainedTypeParameter.ts, 11, 22))
+>Pet : Symbol(Pet, Decl(narrowingConstrainedTypeParameter.ts, 0, 0))
+>pet : Symbol(pet, Decl(narrowingConstrainedTypeParameter.ts, 11, 40))
+>TPet : Symbol(TPet, Decl(narrowingConstrainedTypeParameter.ts, 11, 22))
+>voice : Symbol(voice, Decl(narrowingConstrainedTypeParameter.ts, 11, 50))
+>pet : Symbol(pet, Decl(narrowingConstrainedTypeParameter.ts, 11, 59))
+>TPet : Symbol(TPet, Decl(narrowingConstrainedTypeParameter.ts, 11, 22))
+
+    if (!isPet(pet)) {
+>isPet : Symbol(isPet, Decl(narrowingConstrainedTypeParameter.ts, 5, 1))
+>pet : Symbol(pet, Decl(narrowingConstrainedTypeParameter.ts, 11, 40))
+
+        throw new Error("Expected \"pet\" to be a Pet");
+>Error : Symbol(Error, Decl(lib.d.ts, --, --), Decl(lib.d.ts, --, --))
+    }
+    return voice(pet);
+>voice : Symbol(voice, Decl(narrowingConstrainedTypeParameter.ts, 11, 50))
+>pet : Symbol(pet, Decl(narrowingConstrainedTypeParameter.ts, 11, 40))
+}

--- a/tests/baselines/reference/narrowingConstrainedTypeParameter.types
+++ b/tests/baselines/reference/narrowingConstrainedTypeParameter.types
@@ -1,0 +1,52 @@
+=== tests/cases/compiler/narrowingConstrainedTypeParameter.ts ===
+
+// Repro from #10811
+
+interface Pet {
+>Pet : Pet
+
+    name: string;
+>name : string
+}
+
+function isPet(pet: any): pet is Pet {
+>isPet : (pet: any) => pet is Pet
+>pet : any
+>pet : any
+>Pet : Pet
+
+    return typeof pet.name === "string";
+>typeof pet.name === "string" : boolean
+>typeof pet.name : string
+>pet.name : any
+>pet : any
+>name : any
+>"string" : "string"
+}
+
+export function speak<TPet extends Pet>(pet: TPet, voice: (pet: TPet) => string): string {
+>speak : <TPet extends Pet>(pet: TPet, voice: (pet: TPet) => string) => string
+>TPet : TPet
+>Pet : Pet
+>pet : TPet
+>TPet : TPet
+>voice : (pet: TPet) => string
+>pet : TPet
+>TPet : TPet
+
+    if (!isPet(pet)) {
+>!isPet(pet) : boolean
+>isPet(pet) : boolean
+>isPet : (pet: any) => pet is Pet
+>pet : TPet
+
+        throw new Error("Expected \"pet\" to be a Pet");
+>new Error("Expected \"pet\" to be a Pet") : Error
+>Error : ErrorConstructor
+>"Expected \"pet\" to be a Pet" : string
+    }
+    return voice(pet);
+>voice(pet) : string
+>voice : (pet: TPet) => string
+>pet : TPet
+}

--- a/tests/cases/compiler/narrowingConstrainedTypeParameter.ts
+++ b/tests/cases/compiler/narrowingConstrainedTypeParameter.ts
@@ -1,0 +1,18 @@
+// @strictNullChecks: true
+
+// Repro from #10811
+
+interface Pet {
+    name: string;
+}
+
+function isPet(pet: any): pet is Pet {
+    return typeof pet.name === "string";
+}
+
+export function speak<TPet extends Pet>(pet: TPet, voice: (pet: TPet) => string): string {
+    if (!isPet(pet)) {
+        throw new Error("Expected \"pet\" to be a Pet");
+    }
+    return voice(pet);
+}

